rfcs/20201011-max-pooling-with-global-indices/README.md

```diff
@@ -0,0 +1,66 @@
+# RFC: Max pooling to support global indices as one of the outputs
+
+## Motivation
+
+There are some
+[neural networks for semantic segmentation](Learning_Deconvolution_Network_for_Semantic_Segmentation.pdf)
+which implement the network as two chains: half is a convolutional chain with
+pooling operations in between, another half is a deconvolutional chain with
+**unpooling** operations in between. An unpooling operation does the opposite
+to pooling, it upsamples the image based on the indices obtained from "paired"
+pooling and the output from a previous operation. Currently, oneDNN does not
+give an opportunity to an end user to obtain global indices of local max values.
+The request from PyTorch framework is to add such ability to the max pooling
+operation. The similar request came from ONNX Runtime Team to support their
+feature of running forward and backward pooling in different sub-graphs which
+may execute on different vendor hardware and/or software.
+
+We are not requested to implement an unpooling primitive so far as it is of low
+priority and used in not very popular models (it's not in PyTorch's Model Zoo).
+But even if we were to implement unpooling, we can't rely on workspace memory
+from pooling since in general case the library expect that pooling to be called
+using only oneDNN implementation but not native framework's or user's code. But
+even with the call to oneDNN pooling, using the same workspace would be really
+complicated since the memory format of unpooling may easily differ from pooling
+one, and parameters of pooling and unpooling do not have to coincide - they may
+have different kernel, padding, strides and output sizes.
+
+## Proposal
+
+### Option 1 (recommended) - new algorithm.
+
+Introduce a new max pooling algorithm:
+~~~c
+/* dnnl_types.h */
+
+typedef enum {
+    ...
+    /// Average pooling exclude padding
+    dnnl_pooling_avg_exclude_padding = 0x3ff,
+    /// Max pooling with second output of global indices
+    dnnl_pooling_max_with_global_indices = 0x4ff,
+    ...
+} dnnl_alg_kind_t;
+~~~
+It will instruct the implementation to dump global indices in a second output
+marked as `DNNL_ARG_DST_1`. By introducing a new algorithm we preserve
```

Comment:
Do we know what is the order of the indices should be? Say, we typically use NCHW "natural" order. But what if TF uses NHWC -- would that affect the way a particular index is translated to a number. In other words, could it be that the same conceptual input will produce different "global indices" depending on the framework?

Reply:
Once you asked, now I've understood it's a bit of a problem.

What if we don't do global indices in sense of absolute numbers staying one by one reflecting kernel position but instead return the memory marking the physical position of the biggest value with `1`. This way users can reorder memory the way they need. It will be some sort of what workspace memory now stash but with ability to specify the data type and physical format where the actual global number is a physical offset of the point. When keeping the destination memory and global indices memory in the same memory format, the access can be done via a single offset.

Alternative approach would be to store global dim_t values instead but use memory descriptor format as a way for enumeration or to specify permuted logical axes values to be able to to reflect the physical format of destination memory.

Reply:
> What if we don't do global indices in sense of absolute numbers staying one by one reflecting kernel position but instead return the memory marking the physical position of the biggest value with 1. 

That's an option and interesting idea. But, I am afraid it's too complicated and the corresponding translation will kill all the performance benefits of using oneDNN (I am actually afraid that producing the "proper" global indices will kill the performance benefits anyways).

I think we should first check what the main frameworks do and see if there is a pattern. If there would be too many cases to support, we could consider the alternatives, or just admit that this flavor of max pooling is not worth implementing in oneDNN at all.

> Alternative approach would be to store global dim_t values instead but use memory descriptor format as a way for enumeration or to specify permuted logical axes values to be able to to reflect the physical format of destination memory.

Not sure if this addresses the issue: converting the data in to another layout != converting the array of the indices, but changing the number there (if I understand things correctly). Example:

```
// max pooling: 2x2
A = [0 1]
    [3 2]   // row-major order
global max = [2] ("3" is at position (1,0) --> 2)

but the same array in col-major will give:
global max = [1] // cannot be obtain from "reordering" the global max array above
```

Reply:
I've updated an RFC with Implementation section after our discussion, putting my best English to describe the problem and drawbacks.